src/sql/src/query_model/attribute/relation_type.rs

```diff
@@ -0,0 +1,511 @@
+            ColumnReference(c) => self.cref_type(c, enclosing_box_id),
+            Literal(_, typ) => typ.clone(),
+            CallNullary(func) => func.output_type(),
+            CallUnary { expr, func } => {
+                let input_type = self.expr_type(expr, enclosing_box_id);
+                func.output_type(input_type)
+            }
+            CallBinary { expr1, expr2, func } => {
+                let input1_type = self.expr_type(expr1, enclosing_box_id);
+                let input2_type = self.expr_type(expr2, enclosing_box_id);
+                func.output_type(input1_type, input2_type)
+            }
+            CallVariadic { exprs, func } => {
+                let input_types = exprs
+                    .iter()
+                    .map(|e| self.expr_type(e, enclosing_box_id))
+                    .collect();
+                func.output_type(input_types)
+            }
+            If { then, els, .. } => {
+                let then_type = self.expr_type(then, enclosing_box_id);
+                let else_type = self.expr_type(els, enclosing_box_id);
+                debug_assert!(then_type.scalar_type.base_eq(&else_type.scalar_type));
+                ColumnType {
+                    scalar_type: then_type.scalar_type,
+                    nullable: then_type.nullable || else_type.nullable,
+                }
+            }
+            Aggregate { expr, func, .. } => {
+                let input_type = self.expr_type(expr, enclosing_box_id);
+                func.output_type(input_type)
+            }
+        }
+    }
+
+    /// Looks up a [`ColumnType`] corresponding to the given
+    /// [`ColumnReference`] from the [`RelationType`] attribute
+    /// of the input box, assuming the [`ColumnReference`]
+    /// occurs within the context of the enclosing [`BoxId`].
+    ///
+    /// *Note*: this method should be used only internally in order
+    /// to derive the [`RelationType`] attribute of a [`Model`].
+    fn cref_type(&self, cref: &ColumnReference, enclosing_box_id: BoxId) -> ColumnType {
+        let quantifier = self.get_quantifier(cref.quantifier_id);
+
+        // the type of All and Existential quantifiers is always BOOLEAN NOT NULL
+        if matches!(
+            quantifier.quantifier_type,
+            QuantifierType::Existential | QuantifierType::All
+        ) {
+            return ColumnType {
+                scalar_type: ScalarType::Bool,
+                nullable: false,
+            };
+        }
+
+        let input_box = quantifier.input_box();
+        let column_type = &input_box.attributes.get::<RelationType>()[cref.position];
+
+        let parent_box = self.get_box(quantifier.parent_box);
+        let nullable = if enclosing_box_id == parent_box.id {
+            if quantifier.quantifier_type == QuantifierType::Scalar {
+                // scalar quantifiers can always be NULL if the subquery is empty
+                true
+            } else {
```

Comment:
> I think that it is possible for Q3 to become a scalar quantifier after query rewrites such as merging selects and eliminating join identities.

I am not sure how this will work. In general, I am assuming that correlated `ColumnReference` nodes:

1. Always reference a `Foreach` quantifier.
2. The referenced quantifier precedes the quantifier that denotes the sub-graph where the reference occurs.

Reply:
Looking into the code, I think the way `correlation_info` is currently written is not respecting the above constraint, but I feel that this is essential, otherwise we will often have to derive the dependency graph of correlated quantifiers ad-hoc. If we impose it, we can at least assume that the quantifiers are topologically sorted with respect to the "Qx has a column that appears as correlated in Qy" relation.

Reply:
Hmm, you're probably right that collapsing boxes on the left hand side might only be ok if the RHS was not correlated.

In the [implementation of SelectMerge by Andrés](https://github.com/MaterializeInc/materialize/pull/9689/commits/d13d92fd948c94c95f2db64335897f8e974364a1), it is possible to collapse Box2 into Box0 by removing Q3 and replacing it with the quantifiers in Box2. But that would break the graph because Box6 would still refer to Q3.  (This has been noted in the QGM handover notes so we don't forget.)

I just worry that not fully collapsing boxes would block some query rewrites. Perhaps we could have box collapsing happen only after decorrelation. But this requires more thought.

Reply:
Collapsing boxes is a normalizing transform, so most likely we will need to apply it before applying optimizations that rely on this normalized form and after transforms that break it.

Either way, we need to be careful to conceptualize and represent these things upfront in order to avoid the mistakes from the MIR transform crate. If the result is satisfying we should figure out how to migrate the lessons learned to MIR.